core/src/main/java/com/yahoo/tagchowder/AttributesImpl.java

```diff
@@ -502,51 +574,160 @@ public void setValue(final int index, final String value) {
     ////////////////////////////////////////////////////////////////////
 
     /**
-     * Ensure the internal array's capacity.
+     * Report a bad array index in a manipulator.
      *
-     * @param n The minimum number of attributes that the array must be able to hold.
+     * @param index The index to report.
+     * @exception java.lang.ArrayIndexOutOfBoundsException Always.
      */
-    private void ensureCapacity(final int n) {
-        if (n <= 0) {
-            return;
-        }
-        int max;
-        if (data == null || data.length == 0) {
-            max = 25;
-        } else if (data.length >= n * 5) {
-            return;
-        } else {
-            max = data.length;
-        }
-        while (max < n * 5) {
-            max *= 2;
-        }
+    private void badIndex(final int index) throws ArrayIndexOutOfBoundsException {
+        String msg = "Attempt to modify attribute at illegal index: " + index;
+        throw new ArrayIndexOutOfBoundsException(msg);
+    }
 
-        String[] newData = new String[max];
+    /**
+     * Get the list of indexes to iterate through all entries in attribute class.
+     * @return List of indexes for the attribute entries.
+     */
+    public List<Integer> getIndexes() {
+        ArrayList<Integer> indexes = new ArrayList<>(vectorIndex);
```

Comment:
make is package protected , and return a custom iterator that does not create a new object.